master/buildbot/worker/latent.py

```diff
@@ -183,27 +209,26 @@ def _substantiation_failed(self, failure):
         return self._mail_missing_message(subject, text)
 
     def canStartBuild(self):
-        if self.insubstantiating:
+        if self.insubstantiating or self.retry_timer:
             return False
         return AbstractWorker.canStartBuild(self)
 
     def buildStarted(self, sb):
+        assert sb.isBusy()
         self._clearBuildWaitTimer()
 
     def buildFinished(self, sb):
-        AbstractWorker.buildFinished(self, sb)
-
+        assert not sb.isBusy()
         if not self.building:
             if self.build_wait_timeout == 0:
-                d = self.insubstantiate()
-                # try starting builds for this worker after insubstantiating;
-                # this will cause the worker to re-substantiate immediately if
-                # there are pending build requests.
-                d.addCallback(lambda _:
-                              self.botmaster.maybeStartBuildsForWorker(self.workername))
+                self.insubstantiate()
+                # insubstantiate will automatically retry to create build for this worker
             else:
                 self._setBuildWaitTimer()
 
+        # AbstractWorker.buildFinished() will try to start the next build for that worker
+        AbstractWorker.buildFinished(self, sb)
```

Comment:
This seems redundant. The `buildFinished` method calls `self.botmaster.maybeStartBuildsForWorker()`, but the `insubstantiate` method calls the same method at the end.